### ISSUE TITLE
rclone exits with status 0 after receiving SIGINT/SIGTERM

### ISSUE THREAD

**Issue by hansmi** [rclone/rclone#5437]
#### What is the problem you are having with rclone?

When rclone receives a SIGINT (Ctrl+C) or SIGTERM signal while an atexit function is registered it always exits with status code 0. Unix convention is to exit with a non-zero status code. Often it's `128 + int(signum)`, but at least not zero. Wikipedia writes ([source](https://en.wikipedia.org/wiki/Exit_status#Shell_and_scripts)):

> When a command is terminated by a signal whose number is N, a shell sets the variable $? to a value greater than 128. Most shells use 128+N, while ksh93 uses 256+N.

The non-zero exit status enables scripts to handle commands aborted by signals, e.g. by the user. The responsible code is in `atexit.go:Run` where the exit status is hardcoded: https://github.com/rclone/rclone/blob/e635f4c0be24cdf2272b2c6018939e8469e4c6d5/lib/atexit/atexit.go#L54

I'll be happy to submit a pull request. Would the rclone developers prefer to always use status 2 ("error not otherwise categorised") or the 128+signum convention?

#### What is your rclone version (output from `rclone version`)

```
$ rclone --version
rclone v1.55.1
- os/type: linux
- os/arch: amd64
- go/version: go1.16.3
- go/linking: static
- go/tags: none
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
I did a quick straw poll of a few commands

- `ls`, `tar` exits with 128+SIGNUM
- `rsync` exits with a special error code 20 `received SIGINT, SIGTERM, or SIGHUP (code 20)`

I note that 128+signum is very unix centric

Killing rclone with any other signal will likely exit with 128+signum (though is the done by the shell? not sure).

I think on balance 128+signum is fine, with the proviso that we discover what that does on Windows!

**Comment by hansmi** [comment]
> Killing rclone with any other signal will likely exit with 128+signum (though is the done by the shell? not sure).

If a process handles signals on its own it can implement whatever it wants. That's the case of rclone for SIGINT/SIGTERM on Unix systems.

If a program is terminated by a signal, either by the default handler or because it can't be caught (SIGKILL, etc.), the shell will assign the exit code to 128 + N after it got the signal number from `waitpid` or a related function.

> I think on balance 128+signum is fine, with the proviso that we discover what that does on Windows!

I wasn't aware of this before, but [Windows doesn't have signal numbers](https://github.com/golang/go/issues/30328#issuecomment-465624861). The PR I'm going to push shortly handles this such that on Windows the exit status upon a signal is always 2 (uncategorized error). I cross-built and ran the unit test on Windows.

**Comment by ncw** [comment]
I've merged @hansmi s fix to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.56

Thank you!

**Comment by zackees** [comment]
Hi. I'm facing a similar issue. I'm using rclone in a docker environment. I'm using a memory constrained environment and it looks like the OS is sending a kill signal because rclone is using too much memory. The problem is that rclone is always returning 0 no matter what. I couldn't believe this was the case at first so I did everything possible to get a stack trace and yup, it's always returning 0.

I don't know what the fix is. As reported above, it looks like the exit handler is unconditionally calling exit 0.

https://github.com/rclone/rclone/blob/e635f4c0be24cdf2272b2c6018939e8469e4c6d5/lib/atexit/atexit.go#L54

This is a problem because rclone does not like extremely larges "folders" in s3. For example, we had 80 million files at the root and rclone just runs out of memory and crashes. Even on my local environment.

Always always always return exit code 0.

I've spent the last week reorganizing our s3 bucket so that files go into folders. This has worked great. However, I noticed that the smaller docker instances (500MB Ram) always exit 0 when the OS kills the process. I can see from the logs that the OS is killing the process because it's using too much memory.

This really sucks, because our S3 datalakes are growing and I'd like to get a visual notification that the process has failed to sync so I can take corrective action. At the current rate, this won't be possible. The process exits normally. And then will restart in a few minutes. If I don't catch the issue in the log files then I'll never know that the sync has effectively stopped working. This could go on for months or years and then suddenly when we need the back up THEN we learn that the sync has actually been broken the whole time because one of the directories got big.

Please please please, allow returning non 0 for sigkill for memory overload.

Attached is my script debugging this problem and confirming that 0 is being returned.


```bash
#!/bin/bash

# Enable command tracing
# set -x
# set -e

# DEBUG=${DEBUG:-false}
# DEBUG=true

TRAP_EXIT=1

# Limit the amount of memory rclone can use
alias rclone='rclone --buffer-size 1M --transfers 200 --checkers 100 --max-backlog 1000 --multi-thread-cutoff 50M --use-mmap --vfs-cache-mode minimal --vfs-read-chunk-size 8M --server-side-across-configs --auto-confirm'


# Global variable to store the exit code
SCRIPT_EXIT_CODE=0

# Function to handle non-normal exits
handle_exit() {

    # undo the trap
    trap - EXIT
    local exit_code=$SCRIPT_EXIT_CODE
    echo "Script is exiting with code: $exit_code"

    # Print the stack trace
    echo "Stack trace:"
    local frame=0
    while caller $frame; do
        ((frame++))
    done | while read line; do
        echo "  $line"
    done

    if [ $exit_code -eq 0 ]; then
        echo "Script completed successfully"
        exit 0
    fi

    if [ "$DEBUG" != true ]; then
        echo "Failed"
        exit $exit_code
    fi
    # Wait for user input
    while true; do
        sleep 1
    done
}

# Function to set the exit code
set_exit_code() {
    SCRIPT_EXIT_CODE=$1
}

# Set up trap for non-normal exits
if [ $TRAP_EXIT -eq 1 ]; then
    trap handle_exit EXIT
fi

# Set up trap for errors
trap 'set_exit_code $?' ERR

# Function to check rclone credentials
check_space_credentials() {
    echo "Checking space credentials..."
    # Try to list S3 buckets to verify credentials
    if rclone lsd spaces: --config "$RCLONE_CONFIG"; then
        echo "space credentials are valid."
        return 0
    else
        echo "#####Error: Unable to authenticate with space. Please check your credentials."
        exit 1
    fi
}

check_s3_credentials() {
    echo "Checking S3 credentials..."
    # Try to list S3 buckets to verify credentials
    if rclone lsd s3: --config "$RCLONE_CONFIG" ; then
        echo "S3 credentials are valid."
        return 0
    else
        echo "#####Error: Unable to authenticate with S3. Please check your credentials."
        exit 1
    fi
}

# check that CONFIG exists
if [ ! -f "$RCLONE_CONFIG" ]; then
    echo "Error: rclone config file not found at $RCLONE_CONFIG"
    exit 1
fi

touch /sync.log

check_s3_credentials
check_space_credentials

echo "All credentials checks passed. Proceeding with sync operation."

echo "Running sync from S3 to DigitalOcean Spaces"

# Function to perform sync with retries
perform_sync() {
    rclone sync s3:prod-searches-html spaces:s3-backup-prod-searches-html/prod-searches-html \
                    --progress \
                    --config "$RCLONE_CONFIG" \
                    --stats 60s

    return_value=$?
    if [ $return_value -eq 0 ]; then
        success=true
        echo "Sync completed successfully."
    else
        echo "rclone sync failed with exit code $return_value"
    fi
}

# Function to continuously run perform_sync
continuous_sync() {
    while true; do
        echo "Starting sync operation..."
        perform_sync
        echo "Waiting for 5 minutes before next sync..."
        sleep 300  # Wait for 5 minutes (300 seconds) before next sync
    done
}

# Run the continuous sync
continuous_sync

# Disable command tracing
# set +x
```

```dockerfile
# Use an official Ubuntu base image
FROM ubuntu:24.10

# Install necessary packages
ENV DEBIAN_FRONTEND=noninteractive
RUN apt-get update

RUN apt-get install -y \
    dos2unix \
    awscli \
    curl \
    unzip \
    wget \
    ca-certificates \
    --no-install-recommends

RUN apt-get install -y \
    jq \
    --no-install-recommends

COPY rclone-current-linux-amd64.zip rclone-current-linux-amd64.zip

# Install latest rclone
RUN unzip rclone-current-linux-amd64.zip && \
    cd rclone-*-linux-amd64 && \
    cp rclone /usr/bin/ && \
    chown root:root /usr/bin/rclone && \
    chmod 755 /usr/bin/rclone && \
    cd .. && \
    rm -rf rclone-*-linux-amd64 rclone-current-linux-amd64.zip

# Create rclone config directory
RUN mkdir -p /root/.config/rclone

# Copy rclone config file and AWS credentials
COPY rclone.conf /root/.config/rclone/rclone.conf
COPY .aws /root/.aws

COPY test.sh /test.sh
RUN dos2unix /test.sh
RUN chmod +x /test.sh

# Set AWS config file location
ENV AWS_CONFIG_FILE=/root/.aws/config

# Copy sync script
COPY sync.sh /sync.sh
RUN dos2unix /sync.sh
RUN chmod +x /sync.sh

ENV RCLONE_CONFIG=/root/.config/rclone/rclone.conf

# Run the sync script
CMD ["/sync.sh"]
```
